src/adapter/src/catalog/apply.rs

```diff
@@ -464,127 +461,138 @@ impl CatalogState {
     }
 
     #[instrument(level = "debug")]
-    fn apply_cluster_update(&mut self, cluster: mz_catalog::durable::Cluster, diff: Diff) {
-        apply(
-            &mut self.clusters_by_id,
-            cluster.id,
-            || Cluster {
-                name: cluster.name.clone(),
-                id: cluster.id,
-                bound_objects: BTreeSet::new(),
-                log_indexes: BTreeMap::new(),
-                replica_id_by_name_: BTreeMap::new(),
-                replicas_by_id_: BTreeMap::new(),
-                owner_id: cluster.owner_id,
-                privileges: PrivilegeMap::from_mz_acl_items(cluster.privileges),
-                config: cluster.config.into(),
-            },
-            diff,
-        );
-        apply(
-            &mut self.clusters_by_name,
-            cluster.name,
-            || cluster.id,
-            diff,
-        );
+    fn apply_cluster_update(&mut self, cluster: mz_catalog::durable::Cluster, diff: StateDiff) {
+        apply_inverted(&mut self.clusters_by_name, &cluster.name, cluster.id, diff);
+        apply_catalog_type(&mut self.clusters_by_id, cluster.id, cluster, diff);
     }
 
     #[instrument(level = "debug")]
     fn apply_introspection_source_index_update(
         &mut self,
         introspection_source_index: mz_catalog::durable::IntrospectionSourceIndex,
-        diff: Diff,
+        diff: StateDiff,
     ) {
         let log = BUILTIN_LOG_LOOKUP
             .get(introspection_source_index.name.as_str())
             .expect("missing log");
+        let cluster = self
+            .clusters_by_id
+            .get_mut(&introspection_source_index.cluster_id)
+            .expect("catalog out of sync");
+        apply_inverted(
+            &mut cluster.log_indexes,
+            &log.variant,
+            introspection_source_index.index_id,
+            diff,
+        );
         match diff {
-            1 => {
+            StateDiff::Insert => {
                 self.insert_introspection_source_index(
                     introspection_source_index.cluster_id,
                     log,
                     introspection_source_index.index_id,
                     introspection_source_index.oid,
                 );
             }
-            -1 => {
+            StateDiff::Remove => {
+                self.drop_item(introspection_source_index.index_id);
+            }
+            StateDiff::Update => {
                 self.drop_item(introspection_source_index.index_id);
+                self.insert_introspection_source_index(
+                    introspection_source_index.cluster_id,
+                    log,
+                    introspection_source_index.index_id,
+                    introspection_source_index.oid,
+                );
```

Comment:
This is probably wrong, what about the dependency maps? Look at update_entry and see if we can reuse it.